sdk/dotnet/Pulumi.Automation/Events/EventLogWatcher.cs

```diff
@@ -30,94 +29,74 @@ string logFile
         }
 
-        private async Task PollForEvents()
+        /// Exposed for testing; use Stop instead.
+        internal async Task AwaitPollingTask() 
         {
-            var internalToken = _internalCancellationTokenSource.Token;
-            var externalToken = _externalCancellationToken;
-
-            using var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(internalToken, externalToken);
-
-            var linkedToken = linkedSource.Token;
-
-            while (!linkedToken.IsCancellationRequested)
-            {
-                try
-                {
-                    // NOTE: Waiting before reading so that if ReadEvents throws we will
-                    // wait before attempting it again
-                    await Task.Delay(_pollingIntervalMilliseconds, linkedToken).ConfigureAwait(false);
-                    ReadEvents();
-                }
-                catch (TaskCanceledException)
-                {
-                    break;
-                }
-                catch
-                {
-                    // We do our best effort to keep reading until the polling is cancelled
-                }
-            }
-
-            // Try reading events one final time
             try
             {
-                ReadEvents();
+                await this._pollingTask;
             }
-            catch
+            catch (OperationCanceledException error) when (error.CancellationToken == this._cancellationToken)
             {
-                // ignored
+                // _pollingTask.State == Cancelled
             }
         }
 
-        private void ReadEvents()
+        private async Task PollForEvents(CancellationToken externalCancellationToken)
         {
-            if (!File.Exists(LogFile))
-            {
-                return;
-            }
+            using var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(
+                this._internalCancellationTokenSource.Token,
+                externalCancellationToken);
+            this._cancellationToken = linkedSource.Token;
 
-            using var fs = new FileStream(LogFile, FileMode.Open, FileAccess.Read);
-            var newLength = fs.Length;
+            await ReadEventsOnce();
 
-            if (newLength == _previousLength)
+            // At least one non-interruptible delay to ensure the thread has a chance
+            // to read just-written data.
+            await Task.Delay(_pollingIntervalMilliseconds); 
+
+            while (true)
             {
-                return;
+                await ReadEventsOnce();
+                await Task.Delay(_pollingIntervalMilliseconds, linkedSource.Token);
             }
+        }
 
-            fs.Seek(_previousLength, SeekOrigin.Begin);
-
+        private async Task ReadEventsOnce()
+        {
+            using var fs = new FileStream(LogFile, FileMode.Open, FileAccess.Read)
+            {
+                Position = this._position
+            };
```

Comment:
Should we check that the `LogFile` exists with `File.Exists` before attempting to read? The LogFile may not exist when the `EventLogWatcher` is instantiated because it happens before the we start the `cli` process.

```suggestion
            if (!File.Exists(LogFile))
            {
                return;
            }

            using var fs = new FileStream(LogFile, FileMode.Open, FileAccess.Read)
            {
                Position = this._position
            };
```

Reply:
Reading again, yes you are right, without this there is a race condition. There were a few more things in your original version that might be important let me re-read to make sure they're not lost, if you want to move forward with this version. Sorry I got a bit carried away with this as a C# learning exercise, it would've been better to stick to a smaller diff against the original version.